v2/comments.go

```diff
@@ -229,227 +225,185 @@ func (t Tag) String() string {
-	if s.Scan() != '{' {
-		return 0, 0, s.unexpectedTokenError("JSON array", s.TokenText())
+		out[ident] = append(out[ident], line)
 	}
-	startPos := s.Offset
-	if s.PeekSkipSpaces() == '}' {
-		s.Scan() // consume }
-		return startPos, s.Offset + 1, nil
-	}
-	_, _, err := s.chompJSONFlavoredObjectEntries()
-	if err != nil {
-		return 0, 0, err
-	}
-	if s.Scan() != '}' {
-		return 0, 0, s.unexpectedTokenError("}", s.TokenText())
-	}
-	return startPos, s.Offset + 1, nil
+	return out
 }
 
-func (s *tagKeyScanner) chompJSONFlavoredObjectEntries() (int, int, error) {
-	keyStart, _, err := s.chompJSONFlavoredValue()
-	if err != nil {
-		return 0, 0, err
-	}
-
-	if s.Scan() != ':' {
-		return 0, 0, s.unexpectedTokenError(":", s.TokenText())
-	}
-
-	_, valueEnd, err := s.chompJSONFlavoredValue()
+// ParseTagWithArgs parses a single comment tag with typed args.
+// The tagText must not have a marker or group prefix.
+//
+// The tag may optionally contain function style arguments after the tag name.
+// Arguments may either be a single positional argument or any number of named
+// arguments, but not both. The tag may optionally contain a value following "=".
+// The value text is not parsed and is returned as a string.
+// Argument values may be double-quoted strings, backtick-quoted strings,
+// integers, booleans, or identifiers.
+//
+// Examples:
+//
+//	"tagName"
+//	"tagName=value"
+//	"tagName()"
+//	"tagName(arg)=value"
+//	"tagName("double-quoted")"
+//	"tagName(`backtick-quoted`)"
+//	"tagName(100)"
+//	"tagName(true)"
+//	"name(key1: value1)=value"
+//	"name(key1: value1, key2: value2)"
+//	"name(key1:`string value`)"
+//	"name(key1: 1)"
+//	"name(key1: true)"
+//
+// The tag grammar is:
+//
+// <tag> ::= <name> { "(" { <args> "}" ")" } { "=" <tagValue> }
+// <args> ::= <argValue> | <namedArgs>
+// <namedArgs> ::= <argNameAndValue> { "," <namedArgs> }
+// <argNameAndValue> ::= <identifier> ":" <argValue>
+// <argValue> ::= <identifier> | <string> | <int> | <bool>
+//
+// <identifier> ::= [a-zA-Z_][a-zA-Z0-9_]*
+// <string> ::= [`...` and "..." quoted strings with \\ and \" escaping]
+// <int> ::= [decimal, hex (0x), octal (0o) or binary (0b) notation with optional +/- prefix]
+// <bool> ::= "true" | "false"
+// <tagValue> ::= [all text after the = sign]
+func ParseTagWithArgs(tagText string) (TypedTag, error) {
+	tagText = strings.TrimSpace(tagText)
+	parsed, err := parseTagKey(tagText)
 	if err != nil {
-		return 0, 0, err
+		return TypedTag{}, err
 	}
+	return TypedTag{Name: parsed.name, Args: parsed.args, Value: parsed.value}, nil
+}
 
-	switch s.PeekSkipSpaces() {
-	case ',':
-		s.Scan() // Consume ,
-		_, entriesEnd, err := s.chompJSONFlavoredObjectEntries()
-		if err != nil {
-			return 0, 0, err
+// ExtractAndParseTagWithArgs combines ExtractTags and ParseTagWithArgs.
+func ExtractAndParseTagWithArgs(marker string, group *string, lines []string) (map[TagIdentifier][]TypedTag, error) {
```

Comment:
cute composition, but how much of it needs to be public API?

Reply:
Dropped it.  I kept a `ParseAll` utility. Let me know if even that is too much.